tools/autograd/derivatives.yaml

```diff
@@ -100,11 +100,11 @@
   self: grad * -((-self * self + 1).rsqrt())
 
 - name: add(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor
-  self: grad
-  other: maybe_multiply(grad, alpha)
+  self: grad.to(self.scalar_type())
+  other: maybe_multiply(grad, alpha).to(other.scalar_type())
 
 - name: add(Tensor self, Scalar other, Scalar alpha=1) -> Tensor
-  self: grad
+  self: grad.to(self.scalar_type())
```

Comment:
There are probably other gradients that will need to be changed, but getting feedback on the general approach

Reply:
These shouldn't be necessary. The autograd engine already does some casts internally when the computed gradient has a different type from the source Tensor.

Reply:
yep, you shouldn't need these.

Reply:
If I remove the below cast for mul and run: 
`python test_type_promotion.py TestTypePromotion.test_mixed_type_backward`

I get:
`Function MulBackward0 returned an invalid gradient at index 0 - expected type torch.FloatTensor but got torch.DoubleTensor`

Is it not needed for `add` just because of the lack of use of a mixed-type operator in the gradient definition here?

Reply:
Hmm, interesting, I thought this was done in https://github.com/pytorch/pytorch/blob/30e03df63895600d37d77aa82e96b8b6db5842f2/torch/csrc/autograd/engine.cpp#L423-L468, but it looks like this only handles broadcasting and I was misremembering.

Given that's the case, I think what you have is fine and we can add that automated support later.

Reply:
Oh, I guess we didn't actually implement the casting in engine.cpp. You'll need to adjust this:

https://github.com/pytorch/pytorch/blob/30e03df63895600d37d77aa82e96b8b6db5842f2/torch/csrc/autograd/engine.cpp#L454-L459

Something like `grads[i] = grads[i].to(metadata.type())`

I'm not sure if there are any types we should reject. Maybe dense -> sparse?

Reply:
filed a ticket and assigned to myself:
https://github.com/pytorch/pytorch/issues/22826

Reply:
If you decide to adjust derivatives you'll need to make sure that everything that uses TensorIterator is changed appropriately.

Reply:
I'm not sure what you mean, are you just saying that we need to get rid of the existing `self.scalar_type()` calls in derivatives.yaml for TensorIterator based operators if we fix engine.cpp?

Reply:
I'm saying that any op that uses TensorIterator will now allow mixed-type operations, so you'll have to check that its derivative is correct. For example, the derivative for `sub` isn't correct anymore.

I think this is rather fragile and it's better to just update engine, but that's up to you.

Note that all the unary ops (torch.sin, torch.cos) will allow mixed-type operations with the `out=` syntax, but are probably OK because we don't support autograd through `out=` calls. (e.g. `torch.sin(float_tensor, out=double_tensor)`)

Reply:
ah, I guess I'm not exactly sure how autograd works. sub is implemented as `add(self, -other)` and I thought it just passed through to use the modified add autograd?

I see entries for 'sub' in common_method_invocations and I don't have failing tests. Is there an example of something that should fail if autograd is wrong for sub?

Reply:
I get an error with the following code:

```python
import torch
x = torch.randn(10, requires_grad=True, dtype=torch.float)
y = torch.randn(10, dtype=torch.double)
(x - y).sum().backward()
# RuntimeError: Function SubBackward0 returned an invalid gradient at index 0 - expected type torch.FloatTensor but got torch.DoubleTensor
```

Reply:
Thanks, added this test for each operator and made the suggested engine.cpp change

Reply:
In this failing test:

```
    def test_invalid_gradients(self):
        class MyFunction(Function):
            @staticmethod
            def forward(ctx, x):
                return x * 2

            @staticmethod
            def backward(ctx, grad_output):
                return torch.randn(10, dtype=torch.float)

        with self.assertRaisesRegex(RuntimeError, 'expected shape'):
            input = torch.randn(5, 5, dtype=torch.float, requires_grad=True)
            MyFunction.apply(input).sum().backward()
        with self.assertRaisesRegex(RuntimeError, 'expected type'):
            input = torch.randn(10, dtype=torch.double, requires_grad=True)
            MyFunction.apply(input).sum().backward()
```

I assume I should just remove the second check ('expected type'). Do you know if there's anything else I should be trying to capture here though now that we'll handle types in engine.cpp?